aggregate/billing_functions/utils.py

```diff
@@ -0,0 +1,907 @@
+    """
+    Get the start and end times from the cloud function request.
+    """
+    if request:
+        return request.params['start'], request.params['end']
+    return (None, None)
+
+
+def date_range_iterator(
+    start,
+    end,
+    intv=DEFAULT_RANGE_INTERVAL,
+) -> Iterator[tuple[datetime, datetime]]:
+    """
+    Iterate over a range of dates.
+
+    >>> list(date_range_iterator(datetime(2019, 1, 1), datetime(2019, 1, 2)))
+    [(datetime.datetime(2019, 1, 1, 0, 0), datetime.datetime(2019, 1, 2, 0, 0))]
+
+    >>> list(date_range_iterator(datetime(2019, 1, 1), datetime(2019, 1, 3)))
+    [(datetime.datetime(2019, 1, 1, 0, 0), datetime.datetime(2019, 1, 3, 0, 0))]
+
+    >>> list(date_range_iterator(datetime(2019, 1, 1), datetime(2019, 1, 4)))
+    [(datetime.datetime(2019, 1, 1, 0, 0), datetime.datetime(2019, 1, 3, 0, 0)), (datetime.datetime(2019, 1, 3, 0, 0), datetime.datetime(2019, 1, 4, 0, 0))]
+
+    """  # noqa: E501
+    dt_from = start
+    dt_to = start + intv
+    while dt_to < end:
+        yield (dt_from, dt_to)
+        dt_from += intv
+        dt_to += intv
+
+    dt_to = min(dt_to, end)
+    if dt_from < end:
+        yield (dt_from, end)
+
+
+def get_start_and_end_from_data(data) -> tuple[datetime | None, datetime | None]:
+    """
+    Get the start and end times from the cloud function data.
+    """
+    if data:
+        dates = {}
+        if data.get('attributes'):
+            dates = data.get('attributes', {})
+        elif data.get('start') or data.get('end'):
+            dates = data
+        elif data.get('message'):
+            try:
+                dates = dict(json.loads(data['message']))
+            except ValueError:
+                dates = {}
+
+        logger.info(f'data: {data}, dates: {dates}')
+
+        s_raw = dates.get('start')
+        e_raw = dates.get('end')
+
+        # this should except if the start/end is in an invalid format
+        start = datetime.fromisoformat(s_raw) if s_raw else None
+        end = datetime.fromisoformat(e_raw) if e_raw else None
+
+        return (start, end)
+
+    return (None, None)
+
+
+def process_default_start_and_end(
+    start, end, interval=DEFAULT_RANGE_INTERVAL
+) -> tuple[datetime, datetime]:
+    """
+    Process the start and end times.
+    """
+    if not end and not start:
+        # start of today
+        end = datetime.now()
+        start = end - interval
+    elif not start:
+        start = end - interval
+    elif not end:
+        end = start + interval
+
+    assert isinstance(start, datetime) and isinstance(end, datetime)
+    return start, end
+
+
+def get_date_intervals_for(
+    start, end, interval=DEFAULT_RANGE_INTERVAL
+) -> Iterator[tuple[datetime, datetime]]:
+    """Process start and end times (and get defaults)"""
+    s, e = process_default_start_and_end(start, end)
+    return date_range_iterator(s, e, intv=interval)
+
+
+def read_secret(project_id: str, secret_name: str) -> str | None:
```

Comment:
Don't we have this in `cpg-utils` already?

Reply:
Yeah, we had issues with it. From memory this is slightly refactored. But if they're the same I'll substitute it for cpg-utils one.

Reply:
Yes, but I don't want to pull in cpg-utils (including Hail) for a GCP read secret when it's such a small block of code.

Reply:
Ah fair enough

Reply:
I think it's already a dependency though. Not sure if that's for something else or not. In any case, I'm happy to add it back in and remove the cpg-utils requirement if it's not used for anything else.